src/Aspire.Cli/Commands/DeployCommand.cs

```diff
@@ -0,0 +1,197 @@
+            {
+                env[KnownConfigNames.WaitForDebugger] = "true";
+            }
+
+            var backchannelCompletionSource = new TaskCompletionSource<IAppHostBackchannel>();
+
+            var deployRunOptions = new DotNetCliRunnerInvocationOptions
+            {
+                StandardOutputCallback = deployOutputCollector.AppendOutput,
+                StandardErrorCallback = deployOutputCollector.AppendError,
+                NoLaunchProfile = true
+            };
+
+            var unmatchedTokens = parseResult.UnmatchedTokens.ToArray();
+
+            var pendingRun = _runner.RunAsync(
+                effectiveAppHostProjectFile,
+                false,
+                true,
+                ["--operation", "publish", "--publisher", "default", "--output-path", fullyQualifiedOutputPath, "--deploy", "true", .. unmatchedTokens],
+                env,
+                backchannelCompletionSource,
+                deployRunOptions,
+                cancellationToken);
+
+            // If we use the --wait-for-debugger option we print out the process ID
+            if (waitForDebugger)
+            {
+                _interactionService.DisplayMessage("bug", "Waiting for debugger to attach to app host process");
+            }
+
+            var backchannel = await backchannelCompletionSource.Task.ConfigureAwait(false);
+
+            // Reuse the Publishing activities activity source for the backchannel
+            var deploymentActivities = backchannel.GetPublishingActivitiesAsync(cancellationToken);
+
+            var debugMode = parseResult.GetValue<bool?>("--debug") ?? false;
+
+            var noFailuresReported = debugMode switch
+            {
+                true => await PublishCommand.ProcessPublishingActivitiesAsync(deploymentActivities, cancellationToken),
+                false => await PublishCommand.ProcessAndDisplayPublishingActivitiesAsync(deploymentActivities, cancellationToken),
+            };
+
+            await backchannel.RequestStopAsync(cancellationToken).ConfigureAwait(false);
+            var exitCode = await pendingRun;
+
+            if (exitCode == 0)
+            {
+                _interactionService.DisplaySuccess($"Successfully deployed. Artifacts available at: {fullyQualifiedOutputPath}");
+                return ExitCodeConstants.Success;
+            }
+
+            _interactionService.DisplayLines(deployOutputCollector.GetLines());
+            _interactionService.DisplayError($"Deployment failed with exit code {exitCode}. For more information run with --debug switch.");
+            return ExitCodeConstants.FailedToBuildArtifacts;
+        }
+        catch (OperationCanceledException)
+        {
+            _interactionService.DisplayError("The deployment was canceled.");
+            return ExitCodeConstants.FailedToBuildArtifacts;
+        }
+        catch (ProjectLocatorException ex) when (ex.Message == "Project file is not an Aspire app host project.")
```

Comment:
This catch block looks complicated and duplicate with other commands. Is there a way to refactor this code so it can be shared across commands?

Reply:
I can do a refactor to do more code sharing between the publish and deploy commands since they are closely related to each other (this should cover https://github.com/dotnet/aspire/pull/9792#discussion_r2138886057 as well).

It seems like most of the commands share code around project discovery. In the past, I've used a command base class to encode the `--project` requirement and add subcommands that needed a project to it (see https://github.com/dotnet/aspnetcore/blob/2ee09f5447875667e57a3ea32c7dfef3d3d42e74/src/Tools/dotnet-user-jwts/src/Commands/ProjectCommandLineApplication.cs). It seems like that might be a good pattern for this. Thoughts @mitchdenny?